BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch

```diff
@@ -0,0 +1,258 @@
++        cw.visit(Opcodes.V1_8, Opcodes.ACC_PUBLIC, rawName, null, objectName, new String[]{ eeName });
++
++        // Generate no-args public constructor
++        {
++            MethodVisitor mv = cw.visitMethod(Opcodes.ACC_PUBLIC, "<init>", "()V", null, null);
++            mv.visitVarInsn(Opcodes.ALOAD, 0);
++            mv.visitMethodInsn(Opcodes.INVOKESPECIAL, objectName, "<init>", "()V", false);
++            mv.visitInsn(Opcodes.RETURN);
++            mv.visitMaxs(0, 0);
++            mv.visitEnd();
++        }
++
++        // Generate event executor method
++        {
++            MethodVisitor mv = cw.visitMethod(Opcodes.ACC_PUBLIC, "execute", eeDesc, null, null);
++            mv.visitVarInsn(Opcodes.ALOAD, 1);
++            mv.visitTypeInsn(Opcodes.CHECKCAST, Type.getInternalName(listenerClass));
++            mv.visitVarInsn(Opcodes.ALOAD, 2);
++            mv.visitTypeInsn(Opcodes.CHECKCAST, Type.getInternalName(eventClass));
++            int inv = listenerClass.isInterface() ? Opcodes.INVOKEINTERFACE : Opcodes.INVOKEVIRTUAL;
++            mv.visitMethodInsn(inv, Type.getInternalName(listenerClass), handlerMethod.getName(), Type.getMethodDescriptor(handlerMethod), listenerClass.isInterface());
++            if(handlerMethod.getReturnType() != void.class) {
++                mv.visitInsn(Opcodes.POP);
++            }
++            mv.visitInsn(Opcodes.RETURN);
++            mv.visitMaxs(0, 0);
++            mv.visitEnd();
++        }
++
++        cw.visitEnd();
++        return cw.toByteArray();
++    }
++
++    private static Class<?> defineClass(ClassLoader loader, String name, byte[] data) {
++        try {
++            return (Class<?>) defineClassMethod.invoke(loader, name, data, 0, data.length);
++        } catch (Exception e) {
++            throw new RuntimeException("Failed to define class '" + name + "'", e);
++        }
++    }
++
++    static {
++        try {
++            defineClassMethod = ClassLoader.class.getDeclaredMethod("defineClass", String.class, byte[].class, int.class, int.class);
++            defineClassMethod.setAccessible(true);
++        } catch (Exception e) {
++            throw new Error("Failed to get ClassLoader#defineClass method", e);
++        }
++    }
++}
+\ No newline at end of file
+diff --git a/event/src/main/java/io/github/waterfallmc/waterfall/impl/event/EventExecutor.java b/event/src/main/java/io/github/waterfallmc/waterfall/impl/event/EventExecutor.java
+new file mode 100644
+index 00000000..9f8aac30
+--- /dev/null
++++ b/event/src/main/java/io/github/waterfallmc/waterfall/impl/event/EventExecutor.java
+@@ -0,0 +1,20 @@
++package io.github.waterfallmc.waterfall.impl.event;
++
++import java.lang.reflect.Method;
++import java.lang.reflect.Modifier;
++
++/**
++ * @author Mark Vainomaa
++ */
++public interface EventExecutor {
++    void execute(Object listener, Object event);
++
++    static EventExecutor create(Method method) {
++        if (Modifier.isPublic(method.getModifiers())) {
```

Comment:
does magicaccessor class still exist? if yes, we could do nonpublic calls also with asm

Reply:
Can't access magic stuff past Java 9 iirc, wouldn't risk with that.

Reply:
Actually I don't think the Modifier will ever be nonpublic, as on register it only registers public methods or sth like that

Reply:
Will check it later and make changes.

Reply:
register calls findHandlers, which in turn calls `listener.getClass().getDeclaredMethods()`, so we still end up with non-public methods as I recalled correctly.

Reply:
bungee does call getDeclaredMethods but does never call setAccessible(true), so it would fail calling a nonpublic eventhandler